src/engraving/dom/edit.cpp

```diff
@@ -3580,6 +3558,130 @@ void Score::deleteAnnotationsFromRange(Segment* s1, Segment* s2, track_idx_t tra
+    const auto foundDeselected = [&](const DurationElement* deselectedElement) {
+        const std::vector<Rest*> rests = setRests(restStartTick, track, restDuration, /*useDots*/ !currentTuplet, currentTuplet);
+        crsForSelection.insert(crsForSelection.end(), rests.begin(), rests.end());
+        restStartTick = deselectedElement->tick() + deselectedElement->ticks();
+        restDuration = Fraction();
+    };
+
+    for (Segment* s = startSeg; s && (s->tick() < endTick); s = s->next1()) {
+        EngravingItem* e = s->element(track);
+        if (!e) {
+            continue;
+        }
+
+        if (s->isBreathType()) {
+            deleteItem(e);
+            continue;
+        }
+
+        if (!s->isChordRestType()) {
+            // do not delete TimeSig/KeySig, it doesn't make sense to do it (except on full system)
+            if (!s->isTimeTickType() && !s->isTimeSigType() && !s->isKeySigType()
+                && !s->isType(SegmentType::BarLineType) /*covers all barLine types*/) {
+                undoRemoveElement(e);
+            }
+            continue;
+        }
+
+        if (e->isMeasureRepeat()) {
+            deleteItem(e);
+            continue;
+        }
+
+        ChordRest* cr1 = toChordRest(e);
+
+        // currentTuplet refers to the tuplet that we're currently performing a delete range "inside". It is possible that currentTuplet contains nested
+        // tuplets, in which case we should find the "top-most nested tuplet" (nextTuplet) and:
+        //      1. If it is selected, delete it completely using cmdDeleteTuplet (deletes all elements contained in that tuplet)
+        //      2. If it is not selected, call this method recursively "inside" that tuplet (nextTuplet becomes currentTuplet for the recursive call)
+
+        Tuplet* nextTuplet = nullptr;
+        if (cr1->tuplet() && cr1->tuplet() != currentTuplet) {
+            nextTuplet = cr1->tuplet();
+            while (nextTuplet->tuplet() && nextTuplet->tuplet() != currentTuplet) {
+                nextTuplet = nextTuplet->tuplet();
+            }
+        }
+
+        if (nextTuplet) {
+            //! NOTE: handledTuplets is a slight hack, ideally we'd just skip to the first segment after nextTuplet (tick2segment wasn't reliable for this purpose)
+            if (handledTuplets.find(nextTuplet) != handledTuplets.end()) {
+                continue;
+            }
+            handledTuplets.emplace(nextTuplet);
+
+            if (nextTuplet->selected()) {
+                restDuration += nextTuplet->ticks();
+                cmdDeleteTuplet(nextTuplet, /*replaceWithRest*/ false);
+            } else {
+                const std::vector<ChordRest*> crs = deleteRangeAtTrack(track, cr1->segment(), nextTuplet->tick() + nextTuplet->ticks(),
+                                                                       nextTuplet, handledTuplets);
+                crsForSelection.insert(crsForSelection.end(), crs.begin(), crs.end());
```

Comment:
(To avoid creating small intermediate vectors, you could perhaps pass the one and only `crsForSelection` vector by references to the recursive calls, so that they can modify that same vector directly. But probably wouldn't make a noticeable difference.)